src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,193 @@
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        addToLastPosition(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == 0) {
+            addToFirstPosition(value);
+        } else if (index == size) {
+            addToLastPosition(value);
+        } else {
+            addInsideList(index, value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            addToLastPosition(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexForGet(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = getNodeByIndex(index);
+        removeNode(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = getNodeByValue(object);
+        return removeNode(currentNode);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addToFirstPosition(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        if (head != null) {
+            head.prev = newNode;
+        } else {
+            tail = newNode;
+        }
+        head = newNode;
+        size++;
+    }
+
+    private void addToLastPosition(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail != null) {
+            tail.next = newNode;
+        } else {
+            head = newNode;
+        }
+        tail = newNode;
+        size++;
+    }
+
+    private void addInsideList(int index, T value) {
+        Node<T> currentNode = getNodeByIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        addNodeInsideList(newNode, currentNode);
+    }
+
+    private void addNodeInsideList(Node<T> newNode, Node<T> oldNode) {
+        if (oldNode == head) {
+            addToFirstPosition(newNode.value);
+        } else {
+            newNode.prev = oldNode.prev;
+            oldNode.prev.next = newNode;
+            oldNode.prev = newNode;
+            newNode.next = oldNode;
+        }
+        size++;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || (size == 0 ? index > 0 : index > size)) {
+            throw new IndexOutOfBoundsException("incorrect index");
+        }
+    }
+
+    private void checkIndexForGet(int index) {
```

Comment:
```suggestion
    private void checkIndex(int index) {
```

Reply:
it has two different methods for checking, one of them checkIndexForGet() (when get data) other - checkIndexForAdd (when put data).